### PR TITLE
mega update

### PR DESCRIPTION
**Pull request by Stebalien**
* supersedes #4597 but does not entirely replace it (i.e., not fixed).
* supersedes #4152
* punts #4009

Features:

* fixes build for iOS #4524 (@stebalien, @magik6k)
* new DAG interface from ipfs/go-ipld-format (@stebalien).
* (much) faster base58 (@stebalien).
* faster sha256 on some platforms (@stebalien, @lgierth).
* faster blake2b on all platforms (@stebalien, @lgierth).
* (massively) reduce memory usage of peerstore (@stebalien, @whyrusleeping, @Arceliar).
* #4512: fix dialing without reuseport/on Windows (@stebalien). 
* reliably detect DHT capable peers (@stebalien).
* ~correctly validate DHT records (@dirkmc).~ Use new DHT record validators.
* allocate less in the dht (@stebalien).
* (much) faster metrics (@stebalien).
* fix dropped pubsub messages (@vyzo, @richardschneider).
* fix some potential (rare) connection leaks on dial/accept (@FiloSottile).
* reduce excessive memory usage when mass dialing (@whyrusleeping).
* updated badgerdb (@magik6k)

If I'm missing something here, please comment.

To review this PR, please do so commit by commit for your own sanity.

---

Questions:

* [ ] Are we happy with the fix for #4009 (offline dag)? Currently, I just take a blockservice instead and construct an offline dagservice on the fly (which is what `GetOfflineLinkService` did anyways).

### CODE REVIEW

File: namesys/publisher.go [hunk 1]
```diff
@@ -295,12 +304,31 @@ func selectRecord(recs []*pb.IpnsEntry, vals [][]byte) (int, error) {
...
+	// Note: The DHT will actually check the signature so we don't
+	// need to do that here
+
+	// Author in key must match author in record
+	parts := strings.Split(r.Key, "/")
+	pid, err := peer.IDB58Decode(parts[0])
+	if err != nil {
+		return ErrInvalidAuthor
+	}
+	if string(pid) != string(r.Author) {
```
Comment (dirkmc) [inline namesys/publisher.go]: We can remove the string casts here now right

File: test/sharness/t0183-namesys-pubsub.sh [hunk 2]
```diff
@@ -26,9 +26,11 @@ test_expect_success 'check namesys pubsub state' '
     test_cmp expected state2
 '
 
+# These commands are *expected* to fail. We haven't published anything yet.
 test_expect_success 'subscribe nodes to the publisher topic' '
-    ipfsi 1 name resolve /ipns/$PEERID_0 &&
-    ipfsi 2 name resolve /ipns/$PEERID_0
+    ipfsi 1 name resolve /ipns/$PEERID_0;
+    ipfsi 2 name resolve /ipns/$PEERID_0;
+    true
```
Comment (Stebalien) [inline test/sharness/t0183-namesys-pubsub.sh]: @vyzo does this change make sense?

Comment (vyzo) [reply test/sharness/t0183-namesys-pubsub.sh]: hrm, shouldn't we be testing if they succeeded?
that's why i had the `&&`.

Comment (vyzo) [reply test/sharness/t0183-namesys-pubsub.sh]: the assumption is that name resolve returns 0 if it succeeds.

Comment (Stebalien) [reply test/sharness/t0183-namesys-pubsub.sh]: Yeah but shouldn't it *not* succeed? Nothing has been published yet, as far as I know.

Comment (vyzo) [reply test/sharness/t0183-namesys-pubsub.sh]: yeah, good point! these are executed for their side-effect of subscribing to the topic.

Comment (vyzo) [reply test/sharness/t0183-namesys-pubsub.sh]: So yes, change makes sense.

File: core/commands/dht.go [hunk 3]
```diff
@@ -69,9 +69,13 @@ var queryDhtCmd = &cmds.Command{
 		events := make(chan *notif.QueryEvent)
 		ctx := notif.RegisterForQueryEvents(req.Context(), events)
 
-		k := string(b58.Decode(req.Arguments()[0]))
+		id, err := peer.IDB58Decode(req.Arguments()[0])
```
Comment (Stebalien) [inline core/commands/dht.go]: This broke a test or two but it's *strictly* more correct.

File: unixfs/mod/dagmodifier.go [hunk 4]
```diff
@@ -330,7 +330,6 @@ func (dm *DagModifier) modifyDag(n node.Node, offset uint64, data io.Reader) (*c
 				return nil, false, err
 			}
 
-			offset += bs
```
Comment (Stebalien) [inline unixfs/mod/dagmodifier.go]: @whyrusleeping go vet complained. I believe this is the correct fix here but I just wanted to check. Not really related to this PR, just a drive-by fix.

File: blockservice/blockservice.go [hunk 5]
```diff
@@ -110,38 +122,34 @@ func NewSession(ctx context.Context, bs BlockService) *Session {
...
-	if err != nil {
-		return nil, err
+	if err := s.blockstore.Put(o); err != nil {
+		return err
 	}
 
 	if err := s.exchange.HasBlock(o); err != nil {
-		return nil, errors.New("blockservice is closed")
+		// TODO(stebalien): really an error?
+		return errors.New("blockservice is closed")
```
Comment (Stebalien) [inline blockservice/blockservice.go]: @whyrusleeping Should we not just log and continue here?

File: blockservice/blockservice.go [hunk 6]
```diff
@@ -153,18 +161,16 @@ func (s *blockService) AddBlocks(bs []blocks.Block) ([]*cid.Cid, error) {
...
-		return nil, err
+		return err
 	}
 
-	var ks []*cid.Cid
 	for _, o := range toput {
 		if err := s.exchange.HasBlock(o); err != nil {
-			return nil, fmt.Errorf("blockservice is closed (%s)", err)
+			// TODO(stebalien): Should this really *return*?
+			return fmt.Errorf("blockservice is closed (%s)", err)
```
Comment (Stebalien) [inline blockservice/blockservice.go]: Same?

File: core/commands/pin.go [hunk 7]
```diff
@@ -594,7 +596,10 @@ type pinVerifyOpts struct {
 
 func pinVerify(ctx context.Context, n *core.IpfsNode, opts pinVerifyOpts) <-chan interface{} {
 	visited := make(map[string]PinStatus)
-	getLinks := n.DAG.GetOfflineLinkService().GetLinks
+
+	bs := n.Blocks.Blockstore()
+	DAG := dag.NewDAGService(bserv.New(bs, offline.Exchange(bs)))
+	getLinks := dag.GetLinksWithDAG(DAG)
```
Comment (Stebalien) [inline core/commands/pin.go]: This is the new way to do offline stuff. We could also use the context trick if we get tired of this dance.

File: core/corerepo/gc.go [hunk 8]
```diff
@@ -154,7 +154,7 @@ func GarbageCollectAsync(n *core.IpfsNode, ctx context.Context) <-chan gc.Result
 		return out
 	}
 
-	return gc.GC(ctx, n.Blockstore, n.DAG, n.Pinning, roots)
+	return gc.GC(ctx, n.Blockstore, n.Pinning, roots)
```
Comment (Stebalien) [inline core/corerepo/gc.go]: Now constructs a dag from the blockstore.

File: importer/helpers/helpers.go [hunk 9]
```diff
@@ -133,7 +133,11 @@ func (n *UnixfsNode) FileSize() uint64 {
 }
 
 func (n *UnixfsNode) SetPosInfo(offset uint64, fullPath string, stat os.FileInfo) {
-	n.posInfo = &pi.PosInfo{offset, fullPath, stat}
+	n.posInfo = &pi.PosInfo{
+		Offset:   offset,
+		FullPath: fullPath,
+		Stat:     stat,
+	}
```
Comment (Stebalien) [inline importer/helpers/helpers.go]: `go vet` yelled at me :fearful:.

File: merkledag/batch.go [hunk 10]
```diff
@@ -1,99 +0,0 @@
...
-	blocks "gx/ipfs/Qmej7nf81hi2x2tvjRBF3mcp74sQyuDH4VMYDGd1YtXjb2/go-block-format"
-)
-
-// ParallelBatchCommits is the number of batch commits that can be in-flight before blocking.
-// TODO(#4299): Experiment with multiple datastores, storage devices, and CPUs to find
-// the right value/formula.
-var ParallelBatchCommits = runtime.NumCPU() * 2
-
-// Batch is a buffer for batching adds to a dag.
-type Batch struct {
```
Comment (Stebalien) [inline merkledag/batch.go]: Now in go-ipld-format.

File: blockservice/blockservice.go [hunk 11]
```diff
@@ -110,38 +126,34 @@ func NewSession(ctx context.Context, bs BlockService) *Session {
...
+		// TODO(stebalien): really an error?
+		return errors.New("blockservice is closed")
 	}
 
-	return c, nil
+	return nil
 }
 
-func (s *blockService) AddBlocks(bs []blocks.Block) ([]*cid.Cid, error) {
+func (s *blockService) AddBlocks(bs []blocks.Block) error {
```
Comment (vyzo) [inline blockservice/blockservice.go]: are we losing any functionality by dropping the cids from the return type?

File: core/commands/dht.go (hunk 3 above)
Comment (vyzo) [reply core/commands/dht.go]: sounds like faulty tests!

File: core/commands/pin.go (hunk 7 above)
Comment (vyzo) [reply core/commands/pin.go]: this is 3 lines instead of 1; ergonomically it's not an improvement...

File: namesys/publisher.go [hunk 12]
```diff
@@ -322,7 +322,7 @@ func ValidateIpnsRecord(r *record.ValidationRecord) error {
 	if err != nil {
 		return ErrInvalidAuthor
 	}
-	if string(pid) != string(r.Author) {
+	if pid != r.Author {
```
Comment (vyzo) [inline namesys/publisher.go]: clown shoes!

File: blockservice/blockservice.go (hunk 11 above)
Comment (Stebalien) [reply blockservice/blockservice.go]: Not really. We can get the CIDs from the blocks so there's no need to return them here. It just complicates the blockservice (and dagservice which is why I removed this).

File: namesys/publisher.go (hunk 12 above)
Comment (Stebalien) [reply namesys/publisher.go]: :confused:

File: core/commands/dht.go (hunk 3 above)
Comment (Stebalien) [reply core/commands/dht.go]: Yeah. I just wonder if anyone was relying on this in practice. One could theoretically use this to find the peers closest to some base58 encoded thing that's *not* a valid peer ID.

File: core/commands/pin.go (hunk 7 above)
Comment (Stebalien) [reply core/commands/pin.go]: Discussion: https://github.com/ipfs/go-ipfs/issues/4009

So, first, I made implementing `LinkService` optional. If your `NodeGetter` (the get side of the `DAGService`) doesn't implement the `LinkService` optimizations, it doesn't have to implement the interface.

Next, I tried adding a `OfflineNodeGetter` method to the `NodeGetter` interface. However, that really tricky and messy:

1. We realized that we'd probably want `Offline*` methods on every related service (which would kind of suck).
2. Things like bitswap sessions (which I wanted to turn into valid `NodeGetter`s would *also* have to implement this method). I simply couldn't find a clean way to do this.

So, in #4009, I suggested adding an "offline" flag to the context. I *still* think that's the best way to go because I keep running into similar cases where I want to try to do something without touching the network but @kevina reasonably objected that it was a bit magical and easy to screw up.

Hence this solution.

I could add a helper function but, really, I kind of want this to be ugly. I'm not happy with this solution at all and don't want this pattern repeated all over the codebase. For one, it makes the `LinkService` optimization totally useless because we don't *reuse* the DAG (we construct a new one with the blockstore when we do the GC).

Comment (vyzo) [reply core/commands/pin.go]: fair enough.

**Comment by whyrusleeping** [comment]
@Stebalien sharness on OSX is failing.

**Comment by whyrusleeping** [comment]
```
ok 10 - findpeer

expecting success:
  echo "$TEST_DHT_VALUE" | base64 -d | ipfsi 0 dht put "$TEST_DHT_PATH" | sort >putted &&
  [ -s putted ] ||
  test_fsh cat putted

base64: invalid option -- d
Usage:	base64 [-hvD] [-b num] [-i in_file] [-o out_file]
  -h, --help     display this message
  -D, --decode   decodes input
  -b, --break    break encoded string into num character lines
  -i, --input    input file (default: "-" for stdin)
  -o, --output   output file (default: "-" for stdout)
> cat putted

not ok 11 - put with good keys
#
#	  echo "$TEST_DHT_VALUE" | base64 -d | ipfsi 0 dht put "$TEST_DHT_PATH" | sort >putted &&
#	  [ -s putted ] ||
#	  test_fsh cat putted
#
```

**Comment by whyrusleeping** [comment]
```
why@whys-MacBook-Pro ~/g/s/g/i/g/t/sharness> base64 --help
Usage:	base64 [-hvD] [-b num] [-i in_file] [-o out_file]
  -h, --help     display this message
  -D, --decode   decodes input
  -b, --break    break encoded string into num character lines
  -i, --input    input file (default: "-" for stdin)
  -o, --output   output file (default: "-" for stdout)
```

**Comment by Stebalien** [comment]
And this is why we can't have nice, portable things.

---

Also, that's not the only failure. My little legacy command change also caused an error. `repo gc` calls `SetError` after calling `SetOutput(outChan)`. Do you know the *proper* way to error out after setting the output channel?

**Comment by whyrusleeping** [comment]
@Stebalien SetError should still work, under the hood it should be setting the http trailer. Same as if ipfs add fails out halfway through streaming the file

**Comment by Stebalien** [comment]
> SetError should still work, under the hood it should be setting the http trailer. 

I was worried about concurrency but, as long as we call SetError *before* closing the output channel we should be fine.

However, `SetError` *won't* use a trailer unless we set a *fatal* error. Otherwise, it will just do some weird ducktyping stuff.

File: test/ipfs-test-lib.sh [hunk 13]
```diff
@@ -109,3 +109,13 @@ test_seq() {
...
 }
+
+b64decode() {
+    for flag in "-D" "-d"; do
+        if base64 "$flag" 2>/dev/null; then
+            return
+        fi
+    done
+    echo "no compatible base64 command found" >&2
+    return 1
```
Comment (kevina) [inline test/ipfs-test-lib.sh]: I am not sure how I feel about this little hack.  What if one of the base64 supports both options and they do two different things?

Maybe a simpler and more portable solution would be to write our own tool in go and use that, or install someone else's tool in `test/sharness/bin`.

Comment (Stebalien) [reply test/ipfs-test-lib.sh]: Meh, it's a test case. MacOS uses `-D` and everything else uses `-d`. We could install and use a different command but I'd rather not unless you *really* think this hack is going to be a problem.

Comment (kevina) [reply test/ipfs-test-lib.sh]: Could we at least switch based on the OS?  I will be okay with that.

It could be fine, but it could also create a hard to track down bug, several years out if by the very off chance `-D` does something other then decode in a future version.

Comment (Stebalien) [reply test/ipfs-test-lib.sh]: @kevina fair enough. I was trying to be smart but you're right, being smart is probably a bad idea in this case.

**Comment by Stebalien** [comment]
This actually gives us a tiny but nice performance win (better hashing library? maybe?). Transfer of a large file between two local nodes is now reliably ~1% faster and adding with filestore is also a bit faster (badger may be even faster as we won't have that bottleneck).

**Comment by Stebalien** [comment]
@whyrusleeping tests pass. Some of the commands stuff is still flakey but that's an unrelated bug, as far as I can tell.

**Comment by Stebalien** [comment]
(well, once someone reviews it)

File: core/core.go [hunk 14]
```diff
@@ -43,38 +43,38 @@ import (
...
-	connmgr "gx/ipfs/QmainUZ19HogMBjMocvHmgWEqofbpzLGNAir4AEY1KcJey/go-libp2p-connmgr"
-	mafilter "gx/ipfs/QmcrNvQBZkpbT7zLcGyjdPCiCHLVtaYeg4UVUH7XLfWJWy/go-maddr-filter"
-	addrutil "gx/ipfs/Qmcxa9y1KVC51TVicSKomsnunJGSA9UJSuBvJc28JCip4H/go-addr-util"
-	ds "gx/ipfs/QmdHG8MAuARdGHxx4rPQASLcvhz24fzjSQq7AJRAQEorq5/go-datastore"
-	cid "gx/ipfs/QmeSrf6pzut73u6zLQkRFQ3ygt3k6XFT2kjdYP8Tnkwwyg/go-cid"
-	dht "gx/ipfs/QmfUvYQhL2GinafMbPDYz7VFoZv4iiuLuR33aRsPurXGag/go-libp2p-kad-dht"
+	metrics "gx/ipfs/Qmb1QrSXKwGFWgiGEcyac4s5wakJG4yPvCPk49xZHxr5ux/go-libp2p-metrics"
+	cid "gx/ipfs/QmcZfnkapfECQGcLZaf9B79NRg7cRa9EnZh4LSbkCzwNvY/go-cid"
+	dht "gx/ipfs/Qmdm5sm2xHCXNaWdxpjhFeStvSNMRhKQkqpBX7aDcqXtfT/go-libp2p-kad-dht"
+	node "gx/ipfs/Qme5bWv7wtjUNGsK2BNGVUFPKiuxWrsqrtvYwCLRw8YFES/go-ipld-format"
```
Comment (magik6k) [inline core/core.go]: With more things in `go-ipld-format` we may want to change the naming from `node` to something like `ipld` - imo node is a bit confusing for some of the things that are now there.

**Comment by magik6k** [comment]
Looked at the rest of the changes locally, other than the two comments LGTM

File: core/core.go (hunk 14 above)
Comment (magik6k) [reply core/core.go]: If we decide to, I'm fine with thin being in another PR, this one is big enough.

Comment (Stebalien) [reply core/core.go]: Meh. Might as well do this now. Anyways, go makes this pretty easy :).

**Comment by Stebalien** [comment]
Please ignore codeclimate. It loves to punish those who refactor for the sins of others.

File: blockservice/blockservice.go (hunk 5 above)
Comment (whyrusleeping) [reply blockservice/blockservice.go]: unclear. For now lets leave it as is.

Comment (Stebalien) [reply blockservice/blockservice.go]: Fair. Reported at #4623 (and the comment has been updated to point to the issue).

**Comment by Stebalien** [comment]
CI failure: #3970.

File: core/commands/dht.go (hunk 3 above)
Comment (whyrusleeping) [reply core/commands/dht.go]: should probably make sure that this isnt how we do random-walk bootstrapping

Comment (whyrusleeping) [reply core/commands/dht.go]: oh wait, this is in a command. Nvm

File: blockservice/blockservice.go (hunk 11 above)
Comment (whyrusleeping) [reply blockservice/blockservice.go]: I wouldn't call that an enhancement

File: pin/gc/gc.go [hunk 15]
```diff
@@ -33,15 +35,16 @@ type Result struct {
...
 
 	elock := log.EventBegin(ctx, "GC.lockWait")
 	unlocker := bs.GCLock()
 	elock.Done()
 	elock = log.EventBegin(ctx, "GC.locked")
 	emark := log.EventBegin(ctx, "GC.mark")
 
-	ls = ls.GetOfflineLinkService()
+	bsrv := bserv.New(bs, offline.Exchange(bs))
+	ds := dag.NewDAGService(bsrv)
```
Comment (kevina) [inline pin/gc/gc.go]: This was the way we originally did things back before 721df367a27dfa416e1d93df23ca440fc2431f63.  The reason I put it in was not for catching but for the filestore.  With the current filestore it is not technically needed, but with a richer more functional filestore it might be again.  So that the record is straight I will add a comment outside of the review a little later so its not so easily lost.

File: blockservice/blockservice.go (hunk 11 above)
Comment (kevina) [reply blockservice/blockservice.go]: Fair enough.  But it is still possible for something to go wrong while adding the blocks, such as running out of space.  However, since the underlying BlockStore does not indicate which blocks where added there is no way to get this information.  I will make a separate issue for this so we can figure out a proper solution.

File: pin/gc/gc.go (hunk 15 above)
Comment (Stebalien) [reply pin/gc/gc.go]: I'm really starting to think we should just be using a context hint. Anyways, all this will need to be reworked when GC changed to operate over a DAG instead of a blockservice.

**Comment by kevina** [comment]
This was merged before I had a chance to really review it, sorry I was not more on the ball here but I am _not_ happy with the fix for #4009, I'll add my comment to #4009 where it will have a better chance of being seen.

**Comment by Stebalien** [comment]
That wasn't a fix, it was a workable hack to allow us to move forward and agree on a fix (that's why I called it out and asked if it was ok for now).

**Comment by kevina** [comment]
@Stebalien and sorry again I was not on the ball here.  That change did not imminently register until you closed my p.r.
